teps/0147-tekton-artifacts-phase1.md

```diff
@@ -0,0 +1,285 @@
+
+- Achieve SLSA L3 supply chain security through Tekton `PipelineRuns`
+- End-to-end trust of the provenance produced by Tekton Chains for Tekton `PipelineRuns`
+- `Tasks` to emit a list of artifacts of unknown length.
+
+### Requirements
+
+- Enable `Tasks` to be able to surface Artifact metadata as an array.
+
+## Proposal
+
+Define the structure of the Artifact metadata that the `Step` needs to write out. Define the file path and variable replacement syntax for the location of the metadata file.
+
+The following chapters in the proposal describe in details the proposed phased implementation approach:
+
+- [Artifact Provenance Data](#artifact-metadata)
+- [surfacing artifact metadata into the status](#from-step-into-status)
+- [surfacing the artifact metadata into the provenance payload](#from-status-into-provenance)
+
+
+### Artifact Provenance Data
+
+An artifact can be consumed or produced by a `Task`. This means that we will always have two categories at a high level: `inputs` (to indicate what was downloaded) and `outputs` (to indicate what was uploaded). The metadata content could be defined as follows:
+
+```json
+{
+    "inputs":[
+        "name": "artifact Name", # e.g. source code, dependencies
+        "values": [
+            {
+                "uri": string,
+                "digest": string
+            },
+        ]
+    ],
+    "outputs": [
+        "name": "artifact Name", # e.g. images, packages, coverage reports
+        "values": [
+            {
+                "uri": string,
+                "digest": string
+            },
+        ]
+    ]
+}
+```
+The metadata content is written by the `Step` to a file `$(step.artifacts.path)`:
+
+```yaml
+taskSpec:
+  steps:
+    - name: artifact-reporting
+      image: bash:latest
+      script: |
+        #!/usr/bin/env bash
+        # Download some artifacts
+        # Upload some artifacts
+        echo -n "{\"inputs\":[{\"name\":\"artifact1\",\"values\":[{\"url\":\"some-uri\",\"digest\":\"some-digest\"}]}], \"outputs\":[{\"name\":\"artifact2\",\"values\":[{\"url\":\"some-uri\",\"digest\":\"some-digest\"}]}]}" | tee $(step.artifacts.path)
+```
+The path `$(step.artifacts.path)` resolves to `/tekton/steps/<step-name>/artifacts/provenance.json` in the `Step` container at runtime.
+```yaml
+taskSpec:
+  steps:
+    - name: artifact-reporting
+      image: bash:latest
+      script: |
+        #!/usr/bin/env bash
+        # Download some artifacts
+        # Upload some artifacts
+        echo -n "{\"inputs\":[{\"name\":\"artifact1\",\"values\":[{\"url\":\"some-uri\",\"digest\":\"some-digest\"}]}], \"outputs\":[{\"name\":\"artifact2\",\"values\":[{\"url\":\"some-uri\",\"digest\":\"some-digest\"}]}]}" | tee /tekton/steps/artifact-reporting/artifacts/provenance.json
+```
+
+#### Name
+Name is a `Task`/`Step` defined name. It is useful for referencing (phase 2) and UI. The name the artifact under a particular category (`input` or `output`) must be unique.
```

Comment:
If the name must be unique, should the data structure better represent this? e.g.

```json
{
    "inputs": {
        "<artifact1>": {
                "uri": string,
                "digest": string
        },
        "<artifact2>": {
                "uri": string,
                "digest": string
        }
    },
    "outputs": {
        "<artifact1>": {
                "uri": string,
                "digest": string
        },
        "<artifact2>": {
                "uri": string,
                "digest": string
        }
    }
}
```

Reply:
Since we need **artifacts to be an array**, my thinking here was that we want to name the entire array. e.g. when a `ko` Task produces images, we don't need to name every single image (that's not useful), instead we want to group them all under a single name: e.g. `images`. That's because we  know that the Task is going to be producing `image` artifacts.  
```
outputs:
  name: images
  values:
    - uri: img1
      digest: digest1
    - uri: img2
      digest: digest2
```

This will help with referencing (definitely in phase 2) where we want to enable passing artifacts from one Task to another. 

For example, task 1 produces images and task 2 needs to scan them. For that, we need to pass the image metadata (`uri, digest`) to task2. That's where this name would be handy to pass the array of artifacts. 

Another example: If a Task/Step is producing multiple artifacts types (images, packages etc.), this naming would allow them to group it nicely. Which would help with things like readability for the user, UI-Dashboard etc. where there is a natural grouping.